### ISSUE TITLE
Latest Thor install overwrote my hotkeys

### ISSUE THREAD

**Issue by TamarEGranor** [VFPX/Thor#189]
## 📝 Provide detailed reproduction steps (if any)

1. Opened VFP and Thor CFU ran.

### ✔️ Expected result

Updated tools

### ❌ Actual result

Updated tools and my chosen hotkeys have been replaced with somebody else's choices. (For example, GoFish now has F5 and GoTo definition has F6. Those aren't what my fingers know.) It looks like only the ones I already had defined have hotkeys, but they're all wrong.

## 📷 Screenshots
![image](https://github.com/VFPX/Thor/assets/30271640/041f51d0-8740-451c-aaeb-298695d43ad2)

**Comment by Jimrnelson** [comment]
@TamarEGranor 

Tamar --

So sorry, that's extremely annoying.  I hope you have a backup.  (Hmmm, there's a thought ...)

In any case, there are two things about this that I find mind-boggling:

1.  Check For Updates (updating Thor) does not touch the Hot Keys table
2.  Not only do you not recognize those hot key assignments, neither do I!  I can't for the life of me guess where they might have come from.

So I'm going to ask for some info from you to try to track this down:

1. If you still have the original HotKeyDefinitions file, as shown here, please attach it.  It's found in Thor\Tables.  And, if you haven't modified it, what is the timestamp.
2. Do you have the good (prior) version of this file as well?  Please attach it also.
3. Each time you run CFU, there is a separate folder in  Sys(2023)  named Thor.CheckForUpdates* ... In that folder there is an log file, EventLog.Txt.  Please attach it as well.

Am hoping here that there are some hints in these files as to what happened.

It's rare that I do not have the slightest guess on this one.

**Comment by TamarEGranor** [comment]
A few things. 

First, this is not the machine where I had the problem the other day. I'm back in my office and back on my development machine. So, VFP hadn't been opened in about a week and I have CFU set to run weekly. Thus, it ran when I first opened VFP.

Second, I've already fixed all those hotkeys by letting my fingers tell me what they were supposed to be. 

Third, given that those were Fkeys assigned in sequence, I'm guessing it was something that assigned them in sequence.

That said, are you sure it's HotKeyDefinitions.DBF you're looking for? That one has 175 records and seems like it's just a master list of available hotkeys. I think the one you want is ToolHotKeyAssignments. But since I fixed them all after posting, I don't think it will help.

Attaching the event log from the CFU.
[EventLog.txt](https://github.com/VFPX/Thor/files/13126406/EventLog.txt)

Finally, since I have another machine, copying the ToolHotKeyAssignments table on that one, then opening VFP and running CFU. Same thing happened--hot keys were reassigned to what looks like some default series. I'll add the files you requested, but in a separate post, because it's on a different machine.

**Comment by DougHennig** [comment]
I'm seeing something similar. I had Alt+1 as the hotkey for FoxBin2PRG Convert All Binary Files to Text Files and it got changed to Shift-Ctrl-E (or maybe Shift-Alt-E; I changed it back to Alt+1).

**Comment by TamarEGranor** [comment]
Here are the relevant files from my laptop.
[ForJim.zip](https://github.com/VFPX/Thor/files/13126580/ForJim.zip)
[EventLog.txt](https://github.com/VFPX/Thor/files/13126587/EventLog.txt)

I won't change anything on this machine for a while, in case you need something else.

**Comment by Jimrnelson** [comment]
@TamarEGranor @DougHennig 

I'm beginning to have a glimmer of any idea here.

Am I correct that for each of you the list of tools for which you had hot key assignments was unchanged, but the hot key assignments themselves are completely wrong?

If so, I know what failed, and correcting it will not be fun.

**Comment by TamarEGranor** [comment]
I believe that's correct.

**Comment by Jimrnelson** [comment]
@TamarEGranor @DougHennig 

Tamar  and Doug--

Still bewildered here.  I have found what I thought was a problem (not elaborating here) but that would only have affected abouty 10 keys,.

Do you still have a machine, somewhere, where you have not yet done CFU, correct?  If so, two things:

1. Zip the entire Tables folder and send it. (And save it, too)
2. Send a snapshot of the Hot Key Assignments page in Thor Configuration.

Then run CFU and presumably the hot keys got mangled.  Repeat the same two steps.

**Comment by Jimrnelson** [comment]
If you need to force CFU to re-download Thor, try this first:

`StrToFile(' Force CFU ', _screen.cThorFolder + 'Thorversion.txt', 1)`

Running CFU after will (I hope) force Thor to re-download.


I believe that this problem is related to the adventures of table HotKeyDefinitions, which should be a constant file.

However, something when wrong in the move to Git (2017), and there has been a somewhat mangled version of the correct file (60 fewer records, for instance) as part of Thor since then.

However, I think that this only affected newly installed versions of Thor; I still have the original (173 records).

That's my working theory, but I cannot verify.  I have never had the mangled version.

Starting this weekend, getting an update of Thor always gets the new table.

**Comment by Jimrnelson** [comment]
After a lot of digging and head scratching, I've learned that the problem here is that I did not understand the relationship between three Thor internal tables related to hot key definitions and assignments.  For one thing, I thought there were only two ...  These were tables maintained in Doug's code and, unfortunately, my understanding of them was completely wrong.

First of all, there's yet another Thor update which will prevent this problem from happening to anybody else who has not done a Thor update in the last few days.

Second, to correct the mess this created for each of you, you'll need to find a backup copy of the Tables folder from before Sunday (If you installed Quick Access Menu, there should be a backup folder in the Tables folder)

Restore these two tables from the backup HotKeyDefinitions and ToolHotKeyAssignments

That should put your hot keys back.

Unfortunately, and I say this with great embarrassment, this means that all the hot key assignments so touted in my sessions using the Quick Access Menu, simply do not work for anybody who already had hot key assignments.

**Comment by Steingo** [comment]
Glad to see this thread as I ran into the same issue when I got back on my dev system.  Was able to restore from backup and get my original hotkeys back.  Reran the Thor update this morning and can confirm that this time it preserved my keys.  
However, I'm still getting those weird duplicates on the Quick Access menu:
![thorqa](https://github.com/VFPX/Thor/assets/38174426/2708649e-5e2e-488e-83cb-7a08e7819477)
